chanfitness/chanevent.go

```diff
@@ -0,0 +1,150 @@
+
+	// now is expected to return the current time. It is supplied as an
+	// external function to enable deterministic unit tests.
+	now func() time.Time
+
+	// openedAt tracks the first time this channel was seen, this is not
+	// necessarily the time that it confirmed on chain because channel events
+	// are not persisted at present.
+	openedAt time.Time
+
+	// closedAt is the time that the channel was closed.
+	closedAt time.Time
+}
+
+// add appends an event with the given type and current time to the event log.
+// If the channel is already closed, it returns early without adding an event.
+func (e *chanEventLog) add(eventType eventType) {
+	if !e.closedAt.IsZero() {
+		return
+	}
+
+	e.events = append(e.events, &channelEvent{
+		timestamp: e.now(),
+		eventType: eventType,
+	})
+
+	log.Debugf("Channel %v recording event: %v", e.id, eventType)
+}
+
+// uptime returns total uptime of a channel over the period of time specified
+// by startTime and endTime. Uptime is calculated as the period between an online
+// and offline event. Event logs which end on an online event additionally have
+// the period between the online event and the channel close or end of period
+// added to account for the fact that the peer remained online until that point.
+// If a peer is online at the time a chanEventLog is created, an online event is
+// expected to be added to the log to determine start state.
+func (e *chanEventLog) uptime(startTime, endTime time.Time) time.Duration {
+	var uptime time.Duration
+
+	// Sanity check the period provided.
+	if endTime.Before(startTime) {
+		return 0
+	}
+
+	var (
+		wasOnline  bool
+		lastOnline time.Time
+	)
+
+	// Iterate through all the events in the log until we hit one that is after
+	// the end of the period we are calculating uptime for. Events before the
+	// period startTime need to be processed so we can determine the online/
+	// offline state of the peer at the start of the relevant period.
+	for _, event := range e.events {
+		// If event is after the period we are calculating uptime for, we do not
+		// need to process any further events.
+		if event.timestamp.After(endTime) {
+			break
+		}
+
+		switch event.eventType {
+		case peerOnlineEvent:
+			// Log a warning if two adjacent online events have been observed,
+			// this is unexpected behaviour and indicates that there is a bug.
+			if wasOnline {
+				log.Warn("Adjacent peer online events observed")
+			}
+
+			lastOnline = event.timestamp
+			wasOnline = true
+
+		case peerOfflineEvent:
+			wasOnline = false
+		}
+
+		// If our previous online time is before the period we're calculating
+		// uptime for, it must be progressed to the start time of the period.
+		if lastOnline.Before(startTime) {
+			lastOnline = startTime
+		}
+
+		// If event is before the period we are calculating uptime for, or we
+		// are currently in an online state, we do not need to increment uptime.
+		if event.timestamp.Before(startTime) || wasOnline {
```

Comment:
similarly, can remove `wasOnline`, and move the continue to the `peerOnlineEvent` case